File: leetcode/algorithms/228_summary_ranges/main.java
import java.util.ArrayList;
import java.util.List;

class SummaryRanges {
    /**
     * Two Pointers
     *
     * Complexities:
     *   N - Size of `nums`
     *   - Time Complexity: O(N)
     *   - Space Complexity: O(1)
     */
    public List<String> summaryRanges(int[] nums) {
        List<String> result = new ArrayList<>();

        if (nums.length > 0) {
            int num = nums[0];
            List<Integer> numArray = new ArrayList<>();
            numArray.add(num);

            for (int i = 1; i < nums.length; i++) {
                if (nums[i] - num == 1) {
                    numArray.add(nums[i]);
                } else {
                    if (numArray.size() > 1) {
                        result.add(String.valueOf(numArray.get(0)) + "->"
                                + String.valueOf(numArray.get(numArray.size() - 1)));
                    } else {
                        result.add(String.valueOf(numArray.get(0)));
                    }
                    numArray = new ArrayList<Integer>();
                    numArray.add(nums[i]);
                }
                num = nums[i];
            }

            if (numArray.size() > 0) {
                if (numArray.size() > 1) {
                    result.add(String.valueOf(numArray.get(0)) + "->"
                            + String.valueOf(numArray.get(numArray.size() - 1)));
                } else {
                    result.add(String.valueOf(numArray.get(0)));
                }
            }
        }

        return result;
    }


    // Solution
    /**
     * Two Pointers
     *
     * Complexities:
     *   N - Size of `nums`
     *   - Time Complexity: O(N)
     *   - Space Complexity: O(1)
     */
    public List<String> solution(int[] nums) {
        List<String> result = new ArrayList<>();

        if (nums == null || nums.length == 0) {
            return result;
        }

        int n = nums.length;
        int i = 0;

        while (i < n) {
            int start = nums[i];

            while (i + 1 < n && nums[i] + 1 == nums[i + 1]) {
                i++;
            }

            if (start == nums[i]) {
                result.add(String.valueOf(start));
            } else {
                result.add(start + "->" + nums[i]);
            }

            i++;
        }

        return result;
    }
}

File: leetcode/algorithms/228_summary_ranges/main.py
from typing import List


class SummaryRanges:
    """
    # Two Pointers
    #
    # Complexities:
    #   N - Size of `nums`
    #   - Time Complexity: O(N)
    #   - Space Complexity: O(1)
    """
    def summaryRanges(self, nums: List[int]) -> List[str]:
        result = []

        if len(nums) > 0:
            num = nums[0]
            numArray = [num]

            for i in range(1, len(nums)):
                if nums[i] - num == 1:
                    numArray.append(nums[i])
                else:
                    if len(numArray) > 1:
                        result.append(f"{numArray[0]}->{numArray.pop()}")
                    else:
                        result.append(f"{numArray[0]}")
                    numArray = [nums[i]]
                num = nums[i]

            if len(numArray) > 0:
                if len(numArray) > 1:
                    result.append(f"{numArray[0]}->{numArray.pop()}")
                else:
                    result.append(f"{numArray[0]}")

        return result


    # Solution
    """
    # Two Pointers
    #
    # Complexities:
    #   N - Size of `nums`
    #   - Time Complexity: O(N)
    #   - Space Complexity: O(1)
    """
    def solution(self, nums: List[int]) -> List[str]:
        ans = []
        i = 0
        n = len(nums)
        
        while i < n:
            j = i
            while j + 1 < n and nums[j + 1] == nums[j] + 1:
                j += 1
            
            if i == j:
                ans.append(str(nums[i]))
            else:
                ans.append(f"{nums[i]}->{nums[j]}")
                
            i = j + 1
            
        return ans
